repro/ResponseContext.cxx

```diff
@@ -974,12 +981,47 @@ ResponseContext::processCancel(const SipMessage& request)
 }
 
 void
-ResponseContext::processTimerC()
+ResponseContext::updateTimerC(const resip::Data &tid)
 {
-   if (!mRequestContext.mHaveSentFinalResponse)
+   InfoLog(<<"Updating timer C for client transaction " << tid);
+   int timerCSerial = 0;
+   auto i = mTimerCSerial.find(tid);
+   if (i != mTimerCSerial.end())
+   {
+       timerCSerial = ++i->second;
+   }
+   else
    {
-      InfoLog(<<"Canceling client transactions due to timer C.");
-      cancelAllClientTransactions();
+       mTimerCSerial[tid] = timerCSerial;
+   }
+   TimerCMessage *tc = new TimerCMessage(tid, timerCSerial);
+   mRequestContext.getProxy().postTimerC(std::unique_ptr<TimerCMessage>(tc));
+}
+
+void
+ResponseContext::processTimerC(const resip::Data &tid, int serial)
+{
+   bool isLatestTimer = false;
+   auto i = mTimerCSerial.find(tid);
+   if (i != mTimerCSerial.end())
+   {
+       isLatestTimer = (serial == i->second);
+   }
+
+   if (!mRequestContext.mHaveSentFinalResponse && isLatestTimer)
+   {
+      InfoLog(<<"Canceling client transaction " << tid << " due to timer C.");
+      cancelClientTransaction(tid);
+
+      // After initiating client TX cancellation trigger next target processing  
+      auto i = mActiveTransactionMap.find(tid);
+      if (i != mActiveTransactionMap.end()) {
+          // Reconstruct original target request to be able to generate the response
+          auto targetRequest = buildTargetRequest(i->second);
+          std::unique_ptr<resip::SipMessage> response(Helper::makeResponse(*targetRequest, 408));
+          // Pretend like 408 response has been received to trigger next target processing
+          mRequestContext.process(std::move(response));
+      }
```

Comment:
Couldn't think of a better way to continue target processing than by pretending to receive a `408` response.

Are there alternative/better ways to do it?

Reply:
I appologize if I've missed something (but I am still trying to wrap my head around the code), but if timer C pops, that should trigger a CANCEL to the target, which in turn normally should trigger a 487 on the INVITE.
Sure, with bad/rfc2543-clients, you might just get 2xx on the CANCEL, (or worse, timeout), and in that case maybe a "fake" 408 is good.

From: 3261, 16.8

   If timer C should fire, the proxy MUST either reset the timer with
   any value it chooses, or terminate the client transaction.  If the
   client transaction has received a provisional response, the proxy
   MUST generate a CANCEL request matching that transaction.  If the
   client transaction has not received a provisional response, the proxy
   MUST behave as if the transaction received a 408 (Request Timeout)
   response.

Reply:
> I appologize if I've missed something (but I am still trying to wrap my head around the stack/code as a whole)

No need to appologize, I'm in the same boat as you are - just getting started with resip/repro, evaluating it for potential use and probing for potential feature gaps.

Regarding your comment - I agree, it's a valid point. Ideally, there should be a check for the client TX state and `408` should only be generated if no provisional response has been received, otherwise we should expect client to do it's job and handle CANCEL as per RFC.

I've encountered implementation-level problem with the above approach - as far as I can tell, all TX state info is tracked inside `resip/stack`'s `TransactionController`, which is private to `SipStack` class and as such can not be accessed outside the `resip/stack`. That means that currently there is no easy way to check TX state from `repro`. I can think of some ways to work around that, I'm worried that doing so may affect some fundamental architecture elements.

As such, I would really appreciate some guidance from maintainers of this project regarding:
1. the current PR contents
2. analysis of the problem of accessing TX state from repro
3. potential exposure of `SipStack`'s transaction information (if not already possible)

Reply:
1.  So far I like what I'm seeing.  Looks like  you handled the TCSerial logic nicely.  Great job for someone "just getting started". :)
2/3. Peeking into the TX state shouldn't be needed, we see all of the provisional responses in the ResponseContext code anyway.  In fact, we UpdateTimerC when a provisional arrives and bump the TCSerial.  So,  why not just check if the TCSerial value to see if a provisional has arrived or not (you already have access to it in this code)?

Reply:
> we see all of the provisional responses in the ResponseContext code anyway

I didn't look much in this direction as I was reluctant to keep track of TX state in the `ResponseContext` to avoid complicating it further (aux variable init/cleanup + potentially duplicating some of the state tracking logic).

> So, why not just check if the TCSerial value to see if a provisional has arrived or not

Great idea! Will give it a go.

Reply:
Done